src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -59,20 +60,49 @@ public Task<bool> NominateProjectAsync(string projectUniqueName, CancellationTok
         }
 
         public Task<bool> NominateProjectAsync(string projectUniqueName, IVsProjectRestoreInfo projectRestoreInfo, CancellationToken token)
+        {
+            return NominateProjectAsync(projectUniqueName, projectRestoreInfo, null, token);
+        }
+
+        public Task<bool> NominateProjectAsync(string projectUniqueName, IVsProjectRestoreInfo2 projectRestoreInfo, CancellationToken token)
+        {
+            return NominateProjectAsync(projectUniqueName, null, projectRestoreInfo, token);
+        }
+
+        /// <summary>
+        /// This is where the nominate calls for the IVs1 and IVS3 APIs combine. The reason for this method is to avoid duplication and potential issues 
+        /// The issue with this method is that it has some weird custom logging to ensure backward compatibility. It's on the implementer to ensure these calls are correct.
+        /// <param name="projectUniqueName">projectUniqueName</param>
+        /// <param name="projectRestoreInfo">projectRestoreInfo. Can be null</param>
+        /// <param name="projectRestoreInfo2">proectRestoreInfo2. Can be null</param>
+        /// <param name="token"></param>
+        /// <remarks>Only and exactly one of projectRestoreInfos can be null.</remarks>
+        /// <returns>The task that scheduled restore</returns>
+        private Task<bool> NominateProjectAsync(string projectUniqueName, IVsProjectRestoreInfo projectRestoreInfo, IVsProjectRestoreInfo2 projectRestoreInfo2, CancellationToken token)
         {
             if (string.IsNullOrEmpty(projectUniqueName))
             {
                 throw new ArgumentException(Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(projectUniqueName));
             }
 
-            if (projectRestoreInfo == null)
+            if (projectRestoreInfo == null && projectRestoreInfo2 == null)
             {
                 throw new ArgumentNullException(nameof(projectRestoreInfo));
```

Comment:
That should never be the case which is why I call out in the remarks 
`<remarks>Exactly one of projectRestoreInfos has to null.</remarks>`

It's private so I figured it's on the implementer to ensure that the arguments never both have values. 

Open to any suggestions though. I just didn't want to throw in case the implementers somehow screwed up.

Reply:
A suggestion would be to throw an `ArgumentException` (versus `ArgumentNullException`) if and only if both arguments are non-null.

Reply:
But that would break backwards compatibility with the 1st iteration of the nomination API that throws ArgumentNullException when the `projectRestoreInfo` is null. 

I can something like the following: 

```cs

if (projectRestoreInfo == null && projectRestoreInfo2 == null)
{
                throw new ArgumentNullException(nameof(projectRestoreInfo));
}

if (projectRestoreInfo != null && projectRestoreInfo2 != null)
{
                throw new ArgumentException($"Internal error: Both {nameof(projectRestoreInfo)} and {nameof(projectRestoreInfo2)} cannot have values. Please file an issue at NuGet/Home/issues if you see this exception.}");
}
```